crates/trie/trie/src/hashed_cursor/post_state.rs

```diff
@@ -168,170 +211,52 @@ where
-        if let Some((post_state_entry, db_entry)) = post_state_item.zip(db_item) {
-            // If both are not empty, return the smallest of the two
-            // Post state is given precedence if keys are equal
-            Some(if post_state_entry.0 <= db_entry.0 { post_state_entry } else { db_entry })
-        } else {
-            // Return either non-empty entry
-            db_item.or(post_state_item)
-        }
-    }
-}
-
-impl<C> HashedCursor for HashedPostStateStorageCursor<'_, C>
-where
-    C: HashedStorageCursor<Value = U256>,
-{
-    type Value = U256;
-
-    /// Seek the next account storage entry for a given hashed key pair.
-    fn seek(&mut self, subkey: B256) -> Result<Option<(B256, Self::Value)>, DatabaseError> {
-        let entry = self.seek_inner(subkey)?;
-        self.last_slot = entry.as_ref().map(|entry| entry.0);
+        let entry = self.choose_next_entry()?;
+        self.set_last_key(&entry);
         Ok(entry)
     }
-
-    /// Return the next account storage entry for the current account key.
-    fn next(&mut self) -> Result<Option<(B256, Self::Value)>, DatabaseError> {
-        let next = match self.last_slot {
-            Some(last_slot) => {
-                let entry = self.next_inner(last_slot)?;
-                self.last_slot = entry.as_ref().map(|entry| entry.0);
-                entry
-            }
-            // no previous entry was found
-            None => None,
-        };
-        Ok(next)
-    }
 }
 
-impl<C> HashedStorageCursor for HashedPostStateStorageCursor<'_, C>
+/// The cursor to iterate over post state hashed values and corresponding database entries.
+/// It will always give precedence to the data from the post state.
+impl<C, V> HashedStorageCursor for HashedPostStateCursor<'_, C, V>
 where
-    C: HashedStorageCursor<Value = U256>,
+    C: HashedStorageCursor<Value = V>,
+    V: Copy + Clone + std::fmt::Debug,
 {
     /// Returns `true` if the account has no storage entries.
     ///
     /// This function should be called before attempting to call [`HashedCursor::seek`] or
     /// [`HashedCursor::next`].
     fn is_storage_empty(&mut self) -> Result<bool, DatabaseError> {
-        let is_empty = match &self.post_state_cursor {
-            Some(cursor) => {
-                // If the storage has been wiped at any point
-                self.storage_wiped &&
-                    // and the current storage does not contain any non-zero values
-                    cursor.is_empty()
-            }
-            None => self.cursor.is_storage_empty()?,
-        };
-        Ok(is_empty)
+        // Storage is not empty if it has non-zero slots.
+        if self.post_state_cursor.has_any(|(_, value)| value.is_some()) {
+            return Ok(false);
+        }
+
+        // If no non-zero slots in post state, check the database.
+        // Returns true if cursor is None (wiped storage or empty DB).
+        self.cursor.as_mut().map_or(Ok(true), |c| c.is_storage_empty())
     }
 }
```

Comment:
I'd like to see a proptest like was implemented for InMemoryTrieCursor 🙏

Reply:
There are relevant proptests there already:

https://github.com/paradigmxyz/reth/blob/780161a6472952704c5e1cc6d8f56fbda172d660/crates/trie/db/tests/post_state.rs#L178-L209

https://github.com/paradigmxyz/reth/blob/780161a6472952704c5e1cc6d8f56fbda172d660/crates/trie/db/tests/post_state.rs#L446-L490